src/utility/HCISharedMemTransport.cpp

```diff
@@ -0,0 +1,837 @@
+     * - a local pool of buffer to store packets received from the stack
+     */
+    acl_data = false;
+    return;
+}
+
+static bool acl_data_wait(void)
+{
+    /* Wait 10 sec for previous ACL command to be ack'ed by Low Layers
+     * before sending the next one */
+//    for (unsigned long start = millis(); (millis() - start) < 10000;) {
+    while(1) {
+        /* Wait for 10sec max - if not return an error */
+        if (acl_data) { break; }
+    }
+        if (!acl_data) {
+        /*  no event received, timeout occurs */
+         return false;
+    } else {
+        /*  release immmediately, now that M0 runs */
+        acl_data = false;
+        return true;
+    }
+}
+
+static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+{
+    return;
+}
+
+void shci_notify_asynch_evt(void *pdata)
+{
+    /* Need to parse data in future version */
+    shci_user_evt_proc();
+    return;
+}
+
+void shci_register_io_bus(tSHciIO *fops)
+{
+    /* Register IO bus services */
+    fops->Init    = TL_SYS_Init;
+    fops->Send    = TL_SYS_SendCmd;
+}
+
+HCISharedMemTransportClass::~HCISharedMemTransportClass()
+{
+}
+
+int HCISharedMemTransportClass::begin()
+{
+  /* prepare data */
+  _read_index = 0;
+  _write_index = 0;
+  _write_index_initial = 0;
+
+  memset(_rxbuff, 0, sizeof(_rxbuff));
+
+  start_ble_rf();
+
+        /*  Check whether M0 sub-system was started already by
+         *  checking if the system event was already received
+         *  before. If it was not, then go thru all init. */
+        if (!sys_event) {
+            init_debug();
+            /* Take BLE out of reset */
+            stm32wb_reset();
+            /* "C2 unlocking" */
+            transport_init();
+            /*  At this stage, we got the ready event,
+             *  passed thru TL_SYS_EvtReceived */
+
+            WirelessFwInfo_t wireless_info_instance;
+            WirelessFwInfo_t *p_wireless_info = &wireless_info_instance;
+            SHCI_GetWirelessFwInfo(p_wireless_info);
+            /* 
+             * tr_info("WIRELESS COPROCESSOR FW VERSION ID = %d.%d.%d", p_wireless_info->VersionMajor, p_wireless_info->VersionMinor, p_wireless_info->VersionSub);
+             * tr_info("WIRELESS COPROCESSOR FW STACK TYPE = %d", p_wireless_info->StackType);
+             */
+
+            /* "IPM Channel Open Completed" */
+            return 1;
+       } else {
+            /* "Error opening IPM channel" */
+            return 0; /* error */
+       }
+}
+
+void HCISharedMemTransportClass::end()
+{
+  // Nothing to do here
+}
+
+void HCISharedMemTransportClass::wait(unsigned long timeout)
+{
+  for (unsigned long start = millis(); (millis() - start) < timeout;)
+  {
+    if (available())
+    {
+      break;
+    }
+  }
+}
+
+int HCISharedMemTransportClass::available()
+{
+    /* assuming the reset is already achieved,
+     * the LL-only mode is already configured. */
+    if (_read_index != _write_index) {
+        return 1;
+    }
+    if (sys_event) {
```

Comment:
Not sure to understand why this code is here ?  Can you explain ?

Reply:
this is for checking the fifo of the data to read :+1:  
  write index when storing Rx data
  read index when reading the stored Rx data
(both index equal --> nothing to read)

Reply:
I'm speaking about the check on sys_event - why is there a check in sys_event in this function ?